### PR TITLE
Tweaks to export behaviour in 'check'. Closes #3777

### LINKED ISSUE #3777: `export` behaviour in `check`
- [x] https://github.com/turbot/steampipe/issues/3578
- [x] https://github.com/turbot/steampipe/issues/3577
- [x] https://github.com/turbot/steampipe/issues/3773

### PR DESCRIPTION
**Pull request by binaek**

### CODE REVIEW

File: cmd/check.go [hunk 1]
```diff
@@ -121,85 +120,146 @@ func runCheckCmd(cmd *cobra.Command, args []string) {
...
+		error_helpers.FailOnError(sperr.New("cannot execute 'all' with other benchmarks/controls"))
+	}
+
+	// show the status spinner
+	statushooks.Show(ctx)
+
 	// initialise
-	initData := control.NewInitData(ctx)
+	statushooks.SetStatus(ctx, "Initializing...")
+	initData := control.NewInitData(statushooks.DisableStatusHooks(ctx))
```
Comment (kaidaguerre) [inline cmd/check.go:133]: comment to explain status hooks stuff

File: cmd/check.go [hunk 2]
```diff
@@ -121,85 +120,146 @@ func runCheckCmd(cmd *cobra.Command, args []string) {
...
-	generateSnapshot := shouldShare || shouldUpload
-
-	// treat each arg as a separate execution
-	for _, targetName := range args {
-		if error_helpers.IsContextCanceled(ctx) {
-			durations = append(durations, 0)
-			// skip over this arg, since the execution was cancelled
-			// (do not just quit as we want to populate the durations)
+
+	trees, err := getExecutionTrees(ctx, initData, args...)
```
Comment (kaidaguerre) [inline cmd/check.go:150]: comment! what is going on here?

File: cmd/check.go [hunk 3]
```diff
@@ -121,85 +120,146 @@ func runCheckCmd(cmd *cobra.Command, args []string) {
...
+	error_helpers.FailOnError(err)
+
+	// execute controls synchronously (execute returns the number of alarms and errors)
+	for _, namedTree := range trees { // create a context with check status hooks
```
Comment (kaidaguerre) [inline cmd/check.go:154]: comment out of place?

File: cmd/check.go (hunk 2 above)
Comment (kaidaguerre) [reply cmd/check.go:150]: why would we have more than one tree?

File: cmd/check.go [hunk 4]
```diff
@@ -121,85 +120,146 @@ func runCheckCmd(cmd *cobra.Command, args []string) {
...
+	exportMsg, err := initData.ExportManager.DoExport(ctx, namedTree.name, namedTree.tree, exportArgs)
+	if err != nil {
+		return err
 	}
 
 	// print the location where the file is exported if progress=true
 	if len(exportMsg) > 0 && viper.GetBool(constants.ArgProgress) {
 		fmt.Printf("\n")
-		fmt.Println(strings.Join(exportMsg, "\n"))
+		fmt.Println(exportMsg)
```
Comment (kaidaguerre) [inline cmd/check.go:201]: no need to join? PrintLn handles this?

File: cmd/check.go [hunk 5]
```diff
@@ -121,85 +120,146 @@ func runCheckCmd(cmd *cobra.Command, args []string) {
...
+		statushooks.SetStatus(ctx, "Publishing snapshot")
+		return controldisplay.PublishSnapshot(ctx, executionTree, shouldShare)
+	}
+	return nil
+}
+
+// getExecutionTrees returns a list of execution trees with the names of their export targets
+func getExecutionTrees(ctx context.Context, initData *control.InitData, args ...string) ([]*namedExecutionTree, error) {
+	trees := []*namedExecutionTree{}
+	if initData.ExportManager.HasNamedExport(viper.GetStringSlice(constants.ArgExport)) {
```
Comment (kaidaguerre) [inline cmd/check.go:238]: proper explanation needed!

why are there 2 alternative approaches

File: cmd/check.go [hunk 6]
```diff
@@ -121,85 +120,146 @@ func runCheckCmd(cmd *cobra.Command, args []string) {
...
+	if shouldShare || shouldUpload {
+		statushooks.SetStatus(ctx, "Publishing snapshot")
+		return controldisplay.PublishSnapshot(ctx, executionTree, shouldShare)
+	}
+	return nil
+}
+
+// getExecutionTrees returns a list of execution trees with the names of their export targets
+func getExecutionTrees(ctx context.Context, initData *control.InitData, args ...string) ([]*namedExecutionTree, error) {
+	trees := []*namedExecutionTree{}
```
Comment (kaidaguerre) [inline cmd/check.go:237]: `var trees []*namedExecutionTree`

File: cmd/check.go [hunk 7]
```diff
@@ -121,85 +120,146 @@ func runCheckCmd(cmd *cobra.Command, args []string) {
...
+		// create a single merged execution tree from all arguments
+		executionTree, err := controlexecute.NewExecutionTree(ctx, initData.Workspace, initData.Client, initData.ControlFilterWhereClause, args...)
+		if err != nil {
+			return nil, sperr.WrapWithMessage(err, "could not create merged execution tree")
+		}
+		name := fmt.Sprintf("check.%s", initData.Workspace.Mod.ShortName)
+		trees = append(trees, newNamedExecutionTree(name, executionTree))
+	} else {
+		for _, arg := range args {
+			if error_helpers.IsContextCanceled(ctx) {
```
Comment (kaidaguerre) [inline cmd/check.go:248]: do we need this here?

Comment (kaidaguerre) [reply cmd/check.go:248]: how long does tree creation take?

File: pkg/control/controlexecute/execution_tree.go [hunk 8]
```diff
@@ -56,14 +61,33 @@ func NewExecutionTree(ctx context.Context, workspace *workspace.Workspace, clien
...
 	}
 
-	// now identify the root item of the control list
-	rootItem, err := executionTree.getExecutionRootFromArg(arg)
-	if err != nil {
-		return nil, err
-	}
+	var resolvedItem modconfig.ModTreeItem
 
+	// if only one argument is provided, fall back to the original data structure
```
Comment (kaidaguerre) [inline pkg/control/controlexecute/execution_tree.go:66]:
```
// if only one argument is provided, add this as execution root
```

File: pkg/control/controlexecute/execution_tree.go [hunk 9]
```diff
@@ -56,14 +61,33 @@ func NewExecutionTree(ctx context.Context, workspace *workspace.Workspace, clien
...
+	// tree -> root -> item
```
Comment (kaidaguerre) [inline pkg/control/controlexecute/execution_tree.go:67]: `// tree -> root -> item` remove

File: pkg/control/controlexecute/execution_tree.go [hunk 10]
```diff
@@ -56,14 +61,33 @@ func NewExecutionTree(ctx context.Context, workspace *workspace.Workspace, clien
...
+	if len(args) == 1 {
+		resolvedItem, err = executionTree.getExecutionRootFromArg(args[0])
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// for multiple items, use a virtual benchmark as the parent of the items
```
Comment (kaidaguerre) [inline pkg/control/controlexecute/execution_tree.go:74]: let's call it a root benchmark - its not virtual it's really, it exists in the tree and has children

File: pkg/control/controlexecute/result_group.go [hunk 11]
```diff
@@ -249,13 +253,19 @@ func (r *ResultGroup) addDimensionKeys(keys ...string) {
 	sort.Strings(r.DimensionKeys)
 }
 
-func (r *ResultGroup) addDuration(d time.Duration) {
-	r.updateLock.Lock()
-	defer r.updateLock.Unlock()
+// onChildDone is a callback that gets called from the children of this result group when they are done
+func (r *ResultGroup) onChildDone() {
+	new := atomic.AddUint32(&r.childrenComplete, 1)
```
Comment (kaidaguerre) [inline pkg/control/controlexecute/result_group.go:258]: `new`? what is `new`

more descriptive name please!

**Review by kaidaguerre — CHANGES_REQUESTED** [review]

File: cmd/check.go (hunk 7 above)
Comment (binaek) [reply cmd/check.go:248]: @kaidaguerre 

when a `--where` or `--tag` is passed in, we use the service to filter controls and benchmarks.

**Pull request merged**